gobblin-cluster/src/main/java/org/apache/gobblin/cluster/HelixRetriggeringJobCallable.java

```diff
@@ -181,24 +221,48 @@ private void runJobLauncherLoop() throws JobException {
    */
   private void runJobExecutionLauncher() throws JobException {
     long startTime = 0;
+    String jobName = jobProps.getProperty(ConfigurationKeys.JOB_NAME_KEY);
+    Lock jobLock = locks.get(jobName);
+    boolean locked = false;
+
     try {
+      HelixManager planningJobManager = this.taskDriverHelixManager.isPresent()?
+          this.taskDriverHelixManager.get() : this.jobHelixManager;
+
       String builderStr = jobProps.getProperty(GobblinClusterConfigurationKeys.DISTRIBUTED_JOB_LAUNCHER_BUILDER,
           GobblinHelixDistributeJobExecutionLauncher.Builder.class.getName());
 
       // Check if any existing planning job is running
-      String jobName = jobProps.getProperty(ConfigurationKeys.JOB_NAME_KEY);
       Optional<String> planningJobIdFromStore = jobsMapping.getPlanningJobId(jobName);
 
+      // start of critical section to check if a job with same job name is running
+      jobLock.lock();
+      locked = true;
```

Comment:
Should remove locked since it is not atomic with jobLock.lock() and add a try with finally after this to unlock.